subprojects/execution/src/main/java/org/gradle/internal/execution/timeout/impl/DefaultTimeoutHandler.java

```diff
@@ -16,62 +16,117 @@
 
-        private DefaultTimeout(ScheduledFuture<?> timeoutTask, InterruptOnTimeout interrupter) {
-            this.timeoutTask = timeoutTask;
-            this.interrupter = interrupter;
-        }
+        private final Thread thread;
+        private final Duration timeout;
+        private final Describable workUnitDescription;
 
-        @Override
-        public boolean stop() {
-            timeoutTask.cancel(true);
-            return interrupter.interrupted;
-        }
-    }
+        @Nullable
+        private final BuildOperationRef buildOperationRef;
 
-    private static class InterruptOnTimeout implements Runnable {
-        private final Thread thread;
+        private volatile boolean slowStop;
+        private volatile boolean stopped;
         private volatile boolean interrupted;
 
-        private InterruptOnTimeout(Thread thread) {
+        private volatile ScheduledFuture<?> scheduledFuture;
+
+        private DefaultTimeout(Thread thread, Duration timeout, Describable workUnitDescription, @Nullable BuildOperationRef buildOperationRef) {
             this.thread = thread;
+            this.timeout = timeout;
+            this.workUnitDescription = workUnitDescription;
+            this.buildOperationRef = buildOperationRef;
+
+            scheduledFuture = executor.schedule(this::interrupt, timeout.toMillis(), TimeUnit.MILLISECONDS);
+        }
+
+        private void interrupt() {
+            if (!stopped) {
+                interrupted = true;
+                withSetCurrentBuildOperationRef(() -> LOGGER.warn("Requesting stop of " + workUnitDescription.getDisplayName() + " as it has exceeded its configured timeout of " + TimeFormatting.formatDurationTerse(timeout.toMillis()) + "."));
+                thread.interrupt();
+                scheduledFuture = executor.schedule(this::warnIfNotStopped, slowStopCheckFrequency.toMillis(), TimeUnit.MILLISECONDS);
+            }
+        }
+
+        private void withSetCurrentBuildOperationRef(Runnable runnable) {
+            BuildOperationRef previousBuildOperationRef = currentBuildOperationRef.get();
+            try {
+                currentBuildOperationRef.set(this.buildOperationRef);
+                runnable.run();
+            } finally {
+                currentBuildOperationRef.set(previousBuildOperationRef);
+            }
+        }
+
+        private void warnIfNotStopped() {
+            if (!stopped) {
+                slowStop = true;
+                withSetCurrentBuildOperationRef(() -> LOGGER.warn("Timed out " + workUnitDescription.getDisplayName() + " has not yet stopped."));
+                scheduledFuture = executor.schedule(this::warnIfNotStopped, slowStopCheckFrequency.toMillis(), TimeUnit.MILLISECONDS);
+            }
         }
 
         @Override
-        public void run() {
-            interrupted = true;
-            thread.interrupt();
+        public boolean stop() {
+            stopped = true;
```

Comment:
Should we synchronize between this and the timer?

Currently there are potential races with the result and the logging. Previously, there was also a race regarding the result. 

Synchronizing the two threads would make things much simpler to reason about at least.

Reply:
I have done this in a subsequent commit.

Reply:
I would synchronize. I think there's also a race where the timer can interrupt the worker thread after it has gone and starting doing something else (which is maybe what you meant).

Reply:
Sorry, I didn't see the update.